// @krassowski/jupyterlab-lsp/src/_diagnostics.ts
/* tslint:disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * Default level of the severity for diagnostics without severity provided by the language server.
 */
export type DefaultSeverityLevel = 'Error' | 'Warning' | 'Information' | 'Hint';
/**
 * Codes of diagnostics which should not be shown in the panel nor highlighted in the editor.
 */
export type DiagnosticCodesToIgnore = string[];
/**
 * Regular expressions matching messages of diagnostics which should not be shown in the panel nor highlighted in the editor.
 */
export type DiagnosticMessagesToIgnore = string[];
/**
 * Disable this feature. Requires reloading JupyterLab to apply changes.
 */
export type Disable = boolean;

/**
 * LSP Diagnostics (linter results) settings.
 */
export interface CodeDiagnostics {
  defaultSeverity?: DefaultSeverityLevel;
  ignoreCodes?: DiagnosticCodesToIgnore;
  ignoreMessagesPatterns?: DiagnosticMessagesToIgnore;
  disable?: Disable;
  [k: string]: any;
}
