// @krassowski/jupyterlab-lsp/src/_highlights.ts
/* tslint:disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * Number of milliseconds to delay sending out the highlights request to the language server; you can get better responsiveness adjusting this value, but setting it to zero can actually slow it down as the server might get overwhelmed when moving the cursor.
 */
export type DebouncerDelay = number;
/**
 * Whether the highlights should disappear after the focus leaves the editor/cell
 */
export type RemoveHighlightsOnEditorEGCellBlur = boolean;
/**
 * Disable this feature. Requires reloading JupyterLab to apply changes.
 */
export type Disable = boolean;

/**
 * LSP highlights of the variable under the cursor settings.
 */
export interface CodeHighlights {
  debouncerDelay?: DebouncerDelay;
  removeOnBlur?: RemoveHighlightsOnEditorEGCellBlur;
  disable?: Disable;
  [k: string]: any;
}
