// @krassowski/jupyterlab-lsp/src/_hover.ts
/* tslint:disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * Keyboard key which activates the tooltip on hover. The allowed keys are Alt, Control, Shift, and AltGraph. Linux user: Meta key is also supported. Safari users: Meta key is also supported, AltGraph is not supported. To see which physical keys are mapped, visit: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState
 */
export type ModifierKey = 'Alt' | 'Control' | 'Shift' | 'Meta' | 'AltGraph';
/**
 * Number of milliseconds to delay sending out the hover request to the language server; you can get better responsiveness adjusting this value, but setting it to zero can actually slow it down as the server might get overwhelmed when moving the mouse over the code.
 */
export type ThrottlerDelay = number;
/**
 * Up to how many hover responses should be cached at any given time. The cache being is invalidated after any change in the editor.
 */
export type CacheSize = number;
/**
 * Disable this feature. Requires reloading JupyterLab to apply changes.
 */
export type Disable = boolean;

/**
 * LSP Hover over the code tooltip settings.
 */
export interface CodeHover {
  modifierKey?: ModifierKey;
  throttlerDelay?: ThrottlerDelay;
  cacheSize?: CacheSize;
  disable?: Disable;
  [k: string]: any;
}
